contracts/rewards/SubgraphAvailabilityManager.sol

```diff
@@ -0,0 +1,241 @@
+    uint256 private currentNonce;
+
+    /// @notice Time limit for a vote to be valid
+    uint256 public voteTimeLimit;
+
+    /// @notice Array of oracle addresses
+    address[NUM_ORACLES] public oracles;
+
+    /// @notice Mapping of current nonce to subgraph deployment ID to oracle index to timestamp of last deny vote
+    /// currentNonce => subgraphDeploymentId => oracleIndex => timestamp
+    mapping(uint256 => mapping(bytes32 => mapping(uint256 => uint256))) public lastDenyVote;
+
+    /// @notice Mapping of  current nonce to subgraph deployment ID to oracle index to timestamp of last allow vote
+    /// currentNonce => subgraphDeploymentId => oracleIndex => timestamp
+    mapping(uint256 => mapping(bytes32 => mapping(uint256 => uint256))) public lastAllowVote;
+
+    // -- Events --
+
+    /**
+     * @dev Emitted when an oracle is set
+     */
+    event OracleSet(uint256 indexed index, address indexed oracle);
+
+    /**
+     * @dev Emitted when the vote time limit is set
+     */
+    event VoteTimeLimitSet(uint256 voteTimeLimit);
+
+    /**
+     * @dev Emitted when an oracle votes to deny or allow a subgraph
+     */
+    event OracleVote(
+        bytes32 indexed subgraphDeploymentID,
+        bool deny,
+        uint256 indexed oracleIndex,
+        uint256 timestamp
+    );
+
+    // -- Modifiers --
+
+    modifier onlyOracle(uint256 _oracleIndex) {
+        require(_oracleIndex < NUM_ORACLES, "SAM: index out of bounds");
+        require(msg.sender == oracles[_oracleIndex], "SAM: caller must be oracle");
+        _;
+    }
+
+    // -- Constructor --
+
+    /**
+     * @dev Contract constructor
+     * @param _governor Account that can set or remove oracles and set the vote time limit
+     * @param _rewardsManager Address of the RewardsManager contract
+     * @param _executionThreshold Number of votes required to execute a deny or allow call to the RewardsManager
+     * @param _voteTimeLimit Vote time limit in seconds
+     */
+    constructor(
+        address _governor,
+        address _rewardsManager,
+        uint256 _executionThreshold,
+        uint256 _voteTimeLimit,
+        address[NUM_ORACLES] memory _oracles
+    ) {
+        require(_governor != address(0), "SAM: governor must be set");
+        require(_rewardsManager != address(0), "SAM: rewardsManager must be set");
+        require(
+            _executionThreshold >= NUM_ORACLES.div(2).add(1),
+            "SAM: executionThreshold too low"
+        );
+
+        // Oracles should not be address zero
+        for (uint256 i = 0; i < _oracles.length; i++) {
+            require(_oracles[i] != address(0), "SAM: oracle cannot be address zero");
+        }
+
+        Governed._initialize(_governor);
+        rewardsManager = IRewardsManager(_rewardsManager);
+
+        executionThreshold = _executionThreshold;
+        voteTimeLimit = _voteTimeLimit;
+        oracles = _oracles;
```

Comment:
hmm i think we should set the oracles inside the for loop above and emit `OracleSet` for each one. 

Otherwise the subgraph will have a tough time getting the initial list of oracles

Reply:
Nice catch. I guess we should update the subgraph when we're ready to release this, I'll note that.